io/zenoh-transport/src/common/pipeline.rs

```diff
@@ -447,192 +202,462 @@ impl TransmissionPipeline {
+        &mut self,
+        guard: &mut MutexGuard<'_, Option<SerializationBatch>>,
+    ) -> Vec<SerializationBatch> {
+        let mut batches = vec![];
+        // Empty the ring buffer
+        while let Some(mut batch) = self.s_in.s_out_r.pull() {
+            batch.write_len();
+            batches.push(batch);
+        }
+        // Take the current batch
+        if let Some(batch) = guard.take() {
+            batches.push(batch);
+        }
+        batches
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct TransmissionPipelineConf {
+    pub(crate) is_streamed: bool,
+    pub(crate) batch_size: u16,
+    pub(crate) queue_size: [usize; Priority::NUM],
+    pub(crate) backoff: Duration,
+}
+
+impl Default for TransmissionPipelineConf {
+    fn default() -> Self {
+        Self {
+            is_streamed: true,
+            batch_size: u16::MAX,
+            queue_size: [1; Priority::NUM],
+            backoff: Duration::from_micros(1),
+        }
+    }
+}
+
+pub(crate) struct TransmissionPipeline;
+impl TransmissionPipeline {
+    // A MPSC pipeline
+    pub(crate) fn make(
+        config: TransmissionPipelineConf,
+        conduit: &[TransportConduitTx],
+    ) -> (TransmissionPipelineProducer, TransmissionPipelineConsumer) {
+        let mut stage_in = vec![];
+        let mut stage_out = vec![];
+
+        let default_queue = [Priority::default() as usize];
+        let size_iter = if conduit.len() == 1 {
+            default_queue.iter()
+        } else {
+            config.queue_size.iter()
+        };
+
+        // Create the channel for notifying that new batches are in the out ring buffer
+        // This is a MPSC channel
+        let (n_out_w, n_out_r) = bounded(1);
+
+        for (prio, num) in size_iter.enumerate() {
+            // Create the refill ring buffer
+            // This is a SPSC ring buffer
+            let (mut s_ref_w, s_ref_r) = RingBuffer::<SerializationBatch, RBLEN>::new();
+            // Fill the refill ring buffer with batches
+            for _ in 0..*num {
+                assert!(s_ref_w
+                    .push(SerializationBatch::new(
+                        config.batch_size,
+                        config.is_streamed
+                    ))
+                    .is_none());
             }
+            // Create the channel for notifying that new batches are in the refill ring buffer
+            // This is a SPSC channel
+            let (n_ref_w, n_ref_r) = bounded(1);
+
+            // Create the refill ring buffer
+            // This is a SPSC ring buffer
+            let (s_out_w, s_out_r) = RingBuffer::<SerializationBatch, RBLEN>::new();
+
+            // The batch being serialized upon
+            let current = Arc::new(Mutex::new(None));
+
+            // The stage in for this priority
+            let batching = true;
+            // prio == Priority::Control as usize || prio == Priority::RealTime as usize;
+            let bytes = Arc::new(AtomicU16::new(0));
```

Comment:
`bytes` and `backoff` could be combined into an `Arc<(AU16, ABool)>` to become more compact and reduce ptr-jumping

Reply:
`Arc<AtomicU16>` and `Arc<AtomicBool>` are used for signalling between the reading and writing task. Each of the atomic is used then in opposite `directions`. By putting them in the same `Arc` it might happen that the same cache-line would be invalidate by either writes, which would be unfortunate. In any case, it would be worth to investigate if the use of `CachePadded` would be beneficial or not.